### PR TITLE
add nested recycler view to home screen 

### LINKED ISSUE #93: Nested recycler in home

### PR DESCRIPTION
**Pull request by TarekIdrees**

### CODE REVIEW

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder {
+        return when (viewType) {
+            HEALTHY_VIEW_PAGER -> {
+                val view = LayoutInflater.from(parent.context)
+                    .inflate(R.layout.item_view_pager, parent, false)
+                ViewPagerViewHolder(view)
+            }
+            RECENT_FOOD -> {
+                val view = LayoutInflater.from(parent.context)
+                    .inflate(R.layout.item_horizontal_recipes, parent, false)
```
Comment (Sajadio) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt:39]: Use ViewBinding

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+            setupViewPager(viewPager, advices)
+            setUpTransformer(viewPager)
+            viewAll.setOnClickListener {
+                listener.onViewAllButtonClicked()
+            }
+
+        }
+    }
+
+    private fun ItemViewPagerBinding.setupViewPager(
```
Comment (Sajadio) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt]: Don't use the private   extension  function  and don't use it inside the class 
the recommended uses of extension outside the class and call it in any class

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 3]
```diff
@@ -1,134 +1,113 @@
...
+        homeRecyclerAdapter = HomeRecyclerAdapter(this,itemsList)
+        binding.recyclerViewHome.adapter = homeRecyclerAdapter
     }
 
-    private fun setupVerticalRecipeRecyclerView() {
-        verticalRecipeRecyclerAdapter = VerticalRecipeRecyclerAdapter(this)
-        verticalRecipeRecyclerAdapter.submitList(repository.getRecipes())
-        binding.recyclerViewVertical.adapter = verticalRecipeRecyclerAdapter
+    private fun setupViewPager(){
+        val advices = GetHealthAdvicesUseCase(repository)(ADVICES_LIMIT)
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: use invoke function to be more readable

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+                items,
+                newItems,
+                ::areItemsTheSame
+            )
+        )
+        items = newItems
+        diffResult.dispatchUpdatesTo(this)
+    }
+
+    private fun areItemsTheSame(i: Int, i1: Int, homeItems: List<HomeItem<Any>>): Boolean {
```
Comment (Sajadio) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt]: You don't need this function if you don't have BaseNestedRecyclerView

**Review by Sajadio — COMMENTED** [review]
apply this review and will make your code so clean

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/enums/HomeItem.kt
Comment (AmeerAmjed) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/enums/HomeItem.kt:1]: don't use enums as packge can keep in adapter pakage

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 5]
```diff
@@ -1,134 +1,113 @@
...
     private val repository: Repository by lazy {
         RepositoryImpl(DataSourceProvider.getDataSource(requireActivity().application))
     }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+        setup()
+    }
+
+    private fun setup(){
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: remove white space

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 6]
```diff
@@ -1,134 +1,113 @@
...
     }
 
-    override fun onItemClicked(item: Recipe) {
-        val detailsFragment = DetailsFragment.newInstance(item.id)
+    override fun onCardClicked(id: Int) {
+        val detailsFragment = DetailsFragment.newInstance(id)
         navigateToFragment(detailsFragment)
     }
 
+    override fun onChipClicked(chipId: Int) {
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt:94]: separate this into two functions

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 7]
```diff
@@ -1,134 +1,113 @@
...
+        if(chipId == ALL_RECIPES_ID){
+            itemsList = itemsList.take(3).toMutableList()
+            setupFilteredFoodRecycler()
+            homeRecyclerAdapter.submitList(itemsList)
+        }else{
+            var cuisine = ALL_RECIPES_TEXT
+            when(chipId){
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: attention u use when in else (**bareq**)😃

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+        viewPager.clipChildren = false
+        viewPager.getChildAt(0).overScrollMode = RecyclerView.OVER_SCROLL_NEVER
+
+    }
+
+    private fun ItemViewPagerBinding.setUpTransformer(viewPager: ViewPager2) {
+        val transformer = CompositePageTransformer()
+        transformer.addTransformer(MarginPageTransformer(20))
+        transformer.addTransformer { page, position ->
+            val r = 1 - kotlin.math.abs(position)
```
Comment (AmeerAmjed) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt]: move import **abs** to part , don't import same this away .

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 9]
```diff
@@ -137,26 +116,28 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>(), OnItemClickListener<Re
...
-        handler.postDelayed(runnable, 4000)
+    private fun getAllRecipes(): List<Recipe> {
+        return GetAllFoodUseCase(repository)()
     }
-
-    private val runnable = Runnable {
-        viewPager.currentItem = viewPager.currentItem + 1
-    }
-
     companion object {
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: make it private

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+        override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
+            function(oldItemPosition, newItemPosition, newList)
+
+        override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
+            oldList[oldItemPosition] == newList[newItemPosition]
+
+    }
+
+    companion object {
```
Comment (msalarab) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/adapter/HomeRecyclerAdapter.kt]: make it private

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 11]
```diff
@@ -1,134 +1,113 @@
...
-        }
-        viewPager.setPageTransformer(transformer)
+    override fun onViewAllButtonClicked() {
+        val dishesFragment = DishesFragment()
+        navigateToFragment(dishesFragment)
     }
 
-    override fun onItemClicked(item: Recipe) {
-        val detailsFragment = DetailsFragment.newInstance(item.id)
+    override fun onCardClicked(id: Int) {
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: rename function to `onRecipeCardClick`

File: app/src/main/res/layout/item_view_pager.xml [hunk 12]
```diff
@@ -2,17 +2,50 @@
...
             android:id="@+id/view_pager"
             android:layout_width="match_parent"
             android:layout_height="@dimen/zero"
+            android:layout_marginTop="@dimen/spacing_medium"
             android:paddingHorizontal="@dimen/spacing_medium"
             app:layout_constraintDimensionRatio="16:9"
             app:layout_constraintTop_toTopOf="parent" />
 
+        <TextView
+            android:id="@+id/header_title"
```
Comment (AmeerAmjed) [inline app/src/main/res/layout/item_view_pager.xml]: rename id to text_view_header_title

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 13]
```diff
@@ -1,134 +1,113 @@
...
+        val detailsFragment = DetailsFragment.newInstance(id)
         navigateToFragment(detailsFragment)
     }
 
+    override fun onChipClicked(chipId: Int) {
+        if(chipId == ALL_RECIPES_ID){
+            itemsList = itemsList.take(3).toMutableList()
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: please solve redundant

File: app/src/main/res/layout/item_vertical_recipes.xml [hunk 14]
```diff
@@ -1,20 +1,61 @@
...
+    app:cardCornerRadius="@dimen/radius_medium"
+    app:cardElevation="@dimen/zero"
+    app:strokeWidth="@dimen/zero">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+        <androidx.constraintlayout.utils.widget.ImageFilterView
+            android:id="@+id/card_image"
```
Comment (msalarab) [inline app/src/main/res/layout/item_vertical_recipes.xml]: image_(...)

File: app/src/main/res/layout/item_vertical_recipes.xml [hunk 15]
```diff
@@ -1,20 +1,61 @@
...
+            android:layout_height="@dimen/zero"
+            android:scaleType="centerCrop"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintStart_toStartOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            app:round="@dimen/radius_medium"
+            tools:srcCompat="@tools:sample/backgrounds/scenic" />
+
+        <TextView
+            android:id="@+id/recipe_name"
```
Comment (msalarab) [inline app/src/main/res/layout/item_vertical_recipes.xml]: text_(...)

File: app/src/main/res/layout/item_vertical_recipes.xml [hunk 16]
```diff
@@ -1,20 +1,61 @@
...
+            style="@style/TitleTwoLine"
+            app:layout_constraintBottom_toTopOf="@+id/cuisine_name"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toEndOf="@+id/card_image"
+            app:layout_constraintTop_toTopOf="parent"
+            tools:text="@string/spicy_tomato_rice_recipe" />
+
+        <TextView
+            android:id="@+id/cuisine_name"
```
Comment (msalarab) [inline app/src/main/res/layout/item_vertical_recipes.xml]: text_(...)

File: app/src/main/res/layout/item_vertical_recipes.xml [hunk 17]
```diff
@@ -1,20 +1,61 @@
...
+            android:layout_width="@dimen/zero"
+            android:layout_marginEnd="@dimen/spacing_small"
+            android:layout_marginBottom="@dimen/spacing_normal"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toStartOf="@+id/recipe_time"
+            app:layout_constraintStart_toStartOf="@+id/recipe_name"
+            tools:text="@string/indian" />
+
+        <TextView
+            android:id="@+id/recipe_time"
```
Comment (msalarab) [inline app/src/main/res/layout/item_vertical_recipes.xml]: text_(...)

File: app/src/main/res/layout/item_view_pager.xml (hunk 12 above)
Comment (msalarab) [inline app/src/main/res/layout/item_view_pager.xml]: text_(....)

File: app/src/main/res/layout/item_view_pager.xml [hunk 18]
```diff
@@ -2,17 +2,50 @@
...
+            android:layout_marginStart="@dimen/spacing_medium"
+            android:text="@string/recent_food"
+            android:textColor="@color/tertiary_87"
+            app:layout_constraintTop_toBottomOf="@id/view_pager"
+            app:layout_constraintStart_toStartOf="parent"
+            />
+
+
+        <TextView
+            android:id="@+id/view_all"
```
Comment (msalarab) [inline app/src/main/res/layout/item_view_pager.xml]: text_(..)

**Review by msalarab — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt (hunk 6 above)
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt:94]: separate body to separated functions.

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 19]
```diff
@@ -137,26 +116,28 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>(), OnItemClickListener<Re
...
-    override fun onPause() {
-        super.onPause()
-        handler.removeCallbacks(runnable)
-    }
-
-    override fun onResume() {
-        super.onResume()
-        handler.postDelayed(runnable, 4000)
+    private fun getAllRecipes(): List<Recipe> {
+        return GetAllFoodUseCase(repository)()
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: u can use instance of the use case and call it whenever you need

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt [hunk 20]
```diff
@@ -137,26 +116,28 @@ class HomeFragment : BaseFragment<FragmentHomeBinding>(), OnItemClickListener<Re
...
+        const val DESSERT_ID = R.id.chip_desert
+        const val DINNER_ID = R.id.chip_dinner
+        const val SNACKS_ID = R.id.chip_snack
+
+        const val ALL_RECIPES_TEXT = "All Recipes"
+        const val WORLD_BREAKFAST_TEXT=  "World Breakfast"
+        const val DESSERT_TEXT = "Dessert"
+        const val DINNER_TEXT = "Dinner"
+        const val SNACKS_TEXT = "Snacks"
+
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeFragment.kt]: white spaces :(

File: app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeInteractorListener.kt [hunk 21]
```diff
@@ -0,0 +1,10 @@
+package com.cupcake.chickenmasala.ui.fragment.home
+
+import com.google.android.material.chip.ChipGroup
```
Comment (mosayed01) [inline app/src/main/java/com/cupcake/chickenmasala/ui/fragment/home/HomeInteractorListener.kt]: I guess it's unused import

**Review by mosayed01 — CHANGES_REQUESTED** [review]

**Review by mosayed01 — DISMISSED** [review]

**Review by AmeerAmjed — APPROVED** [review]

**Review by AmeerAmjed — APPROVED** [review]

**Pull request merged**